# Cleaning the data

# 1.Libs

In [62]:
## librerias para limpieza de datos 
import pandas as pd # para leer y manipular datos 
import numpy as np # para operaciones matematicas

## libreria para acceder a funciones del sistema 
import sys 
sys.path.append("../")
import src.suport as sp


## librerias para tratamiento de texto
import re


# opciones de pandas para ver todas las columnas 
pd.set_option('display.max_columns', None)


# 2.Carga de datos

In [2]:
df = pd.read_csv("../data/raw/attacks.csv",encoding = "ISO-8859-1")

In [3]:
df.shape

(25723, 24)

In [4]:
df.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2  original order Unnamed: 22 Unnamed: 23  
0    2018.06.25          6303.0         NaN         NaN  
1    2018.06.18          6302.0         NaN         NaN  
2    2018.06.09          6301.0         NaN         NaN  
3    2018.06.08          6300.0         NaN         NaN  
4    2018.06.04          6299.0         NaN         NaN

In [5]:
df.tail()

Case Number Date  Year Type Country Area Location Activity Name Sex   \
25718         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25719         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25720         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25721         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       Age Injury Fatal (Y/N) Time Species  Investigator or Source  pdf  \
25718  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   
25719  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   
25720  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   
25721  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   
25722  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   

      href formula href Case Number.1 Case Number.2  original order  \
25718          NaN  NaN           NaN           NaN             NaN   
25719          NaN  NaN           NaN           NaN             NaN   
25720          NaN  NaN           NaN           NaN             NaN   
25721          NaN  NaN           NaN           NaN             NaN   
25722          NaN  NaN           NaN           NaN             NaN   

      Unnamed: 22 Unnamed: 23  
25718         NaN         NaN  
25719         NaN         NaN  
25720         NaN         NaN  
25721         NaN         NaN  
25722         NaN         NaN

⚠️Nota: Se observan muchos nulos en ciertas filas

# 3.Limpieza 

## 3.1.Limpiezaformato columnas 

In [6]:
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

Algunas columnas tienen espacios en blanco, mayusculas o nombres que no nos gustan. Procedemos a editar ciertos aspecto para que sean mas homogeneos.

In [7]:
lista_columnas = df.columns
# ponemos todo en minusculas 
for x in lista_columnas:
    print(x.lower())

case number
date
year
type
country
area
location
activity
name
sex 
age
injury
fatal (y/n)
time
species 
investigator or source
pdf
href formula
href
case number.1
case number.2
original order
unnamed: 22
unnamed: 23


In [8]:
# en lista comprimida
lista_columnas = [x.lower() for x in lista_columnas]
print(lista_columnas)

['case number', 'date', 'year', 'type', 'country', 'area', 'location', 'activity', 'name', 'sex ', 'age', 'injury', 'fatal (y/n)', 'time', 'species ', 'investigator or source', 'pdf', 'href formula', 'href', 'case number.1', 'case number.2', 'original order', 'unnamed: 22', 'unnamed: 23']


In [9]:
for x in lista_columnas:
    print(x.replace(" ","_"))

case_number
date
year
type
country
area
location
activity
name
sex_
age
injury
fatal_(y/n)
time
species_
investigator_or_source
pdf
href_formula
href
case_number.1
case_number.2
original_order
unnamed:_22
unnamed:_23


In [10]:
# reemplazamos los espacios por guiones bajos 
lista_columnas = [x.replace(" ","_") for x in lista_columnas] 
print(lista_columnas)

['case_number', 'date', 'year', 'type', 'country', 'area', 'location', 'activity', 'name', 'sex_', 'age', 'injury', 'fatal_(y/n)', 'time', 'species_', 'investigator_or_source', 'pdf', 'href_formula', 'href', 'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22', 'unnamed:_23']


In [11]:
# si el ultimo caracter de es un _ lo eliminamos
lista_columnas = []
for x in lista_columnas:
    if x.endswith("_"):
        lista_columnas.append(x[:-1])
    else:
        lista_columnas.append(x)

In [12]:
# si el ultimo caracter de es un _ lo eliminamos
lista_columnas = [x[:-1] if x.endswith("_") else x for x in lista_columnas]
print(lista_columnas)

[]


> Resumen: limpieza de columnas

In [13]:
# minusculas 
df.columns = [x.lower() for x in df.columns]
df.columns

Index(['case number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex ', 'age', 'injury', 'fatal (y/n)', 'time',
       'species ', 'investigator or source', 'pdf', 'href formula', 'href',
       'case number.1', 'case number.2', 'original order', 'unnamed: 22',
       'unnamed: 23'],
      dtype='object')

In [14]:
# reemplazamos los espacios por guiones bajos 
df.columns = [x.replace(" ","_") for x in df.columns] 
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex_', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species_', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

In [15]:
# si el ultimo caracter de es un _ lo eliminamos
df.columns = [x[:-1] if x.endswith("_") else x for x in df.columns]
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

- Comprobacion de resultados 

In [16]:
df.head(3)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid     USA      Hawaii   

                         location  activity             name sex age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe   F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely    F  11   
2                    Habush, Oahu   Surfing      John Denges   M  48   

                                              injury fatal_(y/n)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   

            time      species          investigator_or_source  \
0          18h00  White shark                R. Collier, GSAF   
1  14h00  -15h00          NaN  K.McMurray, TrackingSharks.com   
2          07h45          NaN  K.McMurray, TrackingSharks.com   

                      pdf                                       href_formula  \
0    2018.06.25-Wolfe.pdf  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  2018.06.18-McNeely.pdf  http://sharkattackfile.net/spreadsheets/pdf_di...   
2   2018.06.09-Denges.pdf  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   

  case_number.2  original_order unnamed:_22 unnamed:_23  
0    2018.06.25          6303.0         NaN         NaN  
1    2018.06.18          6302.0         NaN         NaN  
2    2018.06.09          6301.0         NaN         NaN

## 3.2.Limpieza de duplicados

In [17]:
# miramos la dimension de la tabla
df.shape

(25723, 24)

In [18]:
# detectamos los duplicados absolutos 
df.duplicated().sum() 

19411

In [19]:
# Eliminamos aquellas rows que son exactamente iguales 
df.drop_duplicates(inplace=True) 


In [20]:
# confirmamos que se han eliminado 
df.duplicated().sum()

0

In [21]:
# miramos la dimension de la tabla
df.shape

(6312, 24)

## 3.3.Limpieza de Nulos

### 3.3.1.Analisis de nulos 

In [22]:
# miramos la dimension de la tabla
df.shape

(6312, 24)

#### FILAS

In [23]:
#miramos las filas que tienen todos los valores nulos 
df.isnull().sum(axis=1).value_counts()

3     1515
2     1423
4     1202
5     1194
6      540
7      293
8      102
9       26
10       7
22       7
23       2
24       1
Name: count, dtype: int64

In [24]:
# generamos un filtro para visualizar aquellos registros que tienen mas de 22 valores nulos
df[df.isnull().sum(axis=1) >= 22]

case_number date  year type country area location activity name  sex  \
6302            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6303            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6304            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6305            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6306            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6307            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6308            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6309            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8702          NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       age injury fatal_(y/n) time species investigator_or_source  pdf  \
6302   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6303   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6304   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6305   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6306   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6307   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6308   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
6309   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
8702   NaN    NaN         NaN  NaN     NaN                    NaN  NaN   
25722  NaN    NaN         NaN  NaN     NaN                    NaN  NaN   

      href_formula href case_number.1 case_number.2  original_order  \
6302           NaN  NaN           NaN           NaN          6304.0   
6303           NaN  NaN           NaN           NaN          6305.0   
6304           NaN  NaN           NaN           NaN          6306.0   
6305           NaN  NaN           NaN           NaN          6307.0   
6306           NaN  NaN           NaN           NaN          6308.0   
6307           NaN  NaN           NaN           NaN          6309.0   
6308           NaN  NaN           NaN           NaN          6310.0   
6309           NaN  NaN           NaN           NaN             NaN   
8702           NaN  NaN           NaN           NaN             NaN   
25722          NaN  NaN           NaN           NaN             NaN   

      unnamed:_22 unnamed:_23  
6302          NaN         NaN  
6303          NaN         NaN  
6304          NaN         NaN  
6305          NaN         NaN  
6306          NaN         NaN  
6307          NaN         NaN  
6308          NaN         NaN  
6309          NaN         NaN  
8702          NaN         NaN  
25722         NaN         NaN

In [25]:
# sobre escribimos el df con el filtro quedandonos con aquellos registros que tienen menos de 22 valores nulos
df = df[df.isnull().sum(axis=1) < 22]
df.shape

(6302, 24)

In [26]:
# volvamos a hacer un analisis de los valores nulos por filas 
df.isnull().sum(axis=1).value_counts()

3     1515
2     1423
4     1202
5     1194
6      540
7      293
8      102
9       26
10       7
Name: count, dtype: int64

In [27]:
# repetimos un filtro pero esta vez para aquellos registros con 10 casillas nulas para hacer a "ojimetro" un analisis de nuestros datos y sacar insights
df[df.isnull().sum(axis=1) == 10]

case_number                  date    year          type  country area  \
4976    1932.11.09           08-Nov-1932  1932.0  Sea Disaster     CUBA  NaN   
5036  1930.09.12.R                  1930  1930.0       Boating   AZORES  NaN   
6080  1836.07.26.R          1836.07.26.R  1836.0       Invalid    SPAIN  NaN   
6108  1823.12.30.R  Reported 30-Dec-1823  1823.0       Invalid  JAMAICA  NaN   
6122  1808.06.01.R  Reported 01-May-1808  1808.0  Sea Disaster  ANTIGUA  NaN   
6130  1802.04.13.R  Reported Apr-13-1802  1802.0    Unprovoked    INDIA  NaN   
6165  1638.00.00.R         Reported 1638  1638.0    Unprovoked      NaN  NaN   

              location                activity     name  sex  age  \
4976               NaN  Hurricane & Tidal Wave      NaN  NaN  NaN   
5036               NaN                     NaN      NaN  NaN  NaN   
6080               NaN                     NaN      NaN  NaN  NaN   
6108               NaN                     NaN     male  NaN  NaN   
6122  Falmouth Harbour                     NaN      NaN  NaN  NaN   
6130               NaN                     NaN      NaN  NaN  NaN   
6165               NaN                     NaN  sailors    M  NaN   

                                        injury fatal_(y/n) time  \
4976                                       NaN     UNKNOWN  NaN   
5036  No inury to occupants, shark struck boat           N  NaN   
6080     Shark caught, contained human remains         NaN  NaN   
6108              Human remains found in shark         NaN  NaN   
6122            Probable drowning & scavenging         NaN  NaN   
6130                                     FATAL           Y  NaN   
6165                                       NaN     UNKNOWN  NaN   

                                           species  \
4976                                           NaN   
5036                                           NaN   
6080  Shark involvement prior to death unconfirmed   
6108  Shark involvement prior to death unconfirmed   
6122                                           NaN   
6130                                           NaN   
6165                                           NaN   

                 investigator_or_source                            pdf  \
4976          Barrier Miner, 11/14/1932  1932.11.09-Hurricane-Cuba.pdf   
5036                     C. Moore, GSAF        1930.09.12.R-Azores.pdf   
6080                     C. Moore, GSAF         1836.07.26.R-Spain.pdf   
6108                                NaN       1823.12.30.R-Jamaica.pdf   
6122          London Observer, 5/1/1808       1808.05.01.R-Antigua.pdf   
6130  Evening Post (New York) 4/13/1802         1802.04.13.R-India.pdf   
6165                 Sir Thomas Herbert           1638.00.00.R-Herbert   

                                           href_formula  \
4976  http://sharkattackfile.net/spreadsheets/pdf_di...   
5036  http://sharkattackfile.net/spreadsheets/pdf_di...   
6080  http://sharkattackfile.net/spreadsheets/pdf_di...   
6108  http://sharkattackfile.net/spreadsheets/pdf_di...   
6122  http://sharkattackfile.net/spreadsheets/pdf_di...   
6130  http://sharkattackfile.net/spreadsheets/pdf_di...   
6165  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
4976  http://sharkattackfile.net/spreadsheets/pdf_di...    1932.11.09   
5036  http://sharkattackfile.net/spreadsheets/pdf_di...  1930.09.12.R   
6080  http://sharkattackfile.net/spreadsheets/pdf_di...  1836.07.26.R   
6108  http://sharkattackfile.net/spreadsheets/pdf_di...  1823.12.30.R   
6122  http://sharkattackfile.net/spreadsheets/pdf_di...  1808.05.01.R   
6130  http://sharkattackfile.net/spreadsheets/pdf_di...  1802.04.13.R   
6165  http://sharkattackfile.net/spreadsheets/pdf_di...  1638.00.00.R   

     case_number.2  original_order unnamed:_22 unnamed:_23  
4976    1932.11.09          1327.0         NaN         NaN  
5036  1930.09.12.R          1267.0         NaN         NaN  
6080  183

Analizando la tabla por columnas 
- Se observa como las columnas `case_number`,`case_number.1` y `case_number.2` contienen informacion similar. Concretamente parecen tener informacion referente a fecha y en algunos casos una letra como identificador del caso. 

    > Accion_case_number:
    - Son fuente de informacion para completar informacion referente a fechas como las columnas `date` o `year`
    - Eliminar dos de ellas tras completar la informacion. 

- Se observa como las columnas `injury` y `fatal_(y/n)` podria llegar a retroalimentarse en cas de datos perdidos. Vease el registro `6130`. Ademas la columna `fatal_(y/n)` aparece con un valor `UNKNOWN`, este valor resulta confuso por lo que deberiamos hacer un analisis esos valores. 
    > Accion_fatal_(y/n):
    - Analizar la columna `injury` para evaluar los valores `UNKNOWN`.
    - La columna `injury` podria llegar a completar `fatal_(y/n)`.
    

- Analizamos las columnas `href_formula` y `href` mas en profundidad

Comprobamos si la informacion de las columnas coinciden.

In [28]:
df.loc[6165,"href_formula"]

'http://sharkattackfile.net/spreadsheets/pdf_directory/1638.00.00.R-Herbert'

In [29]:
df.loc[6165,"href"]

'http://sharkattackfile.net/spreadsheets/pdf_directory/1638.00.00.R-Herbert'

Parece que para el registro `6165` funciona. Pero vamos a comprobar una funcion para que compruebe si los datos de varias columnas no coinciden para tener mayor perspectiva

In [66]:
df[df["href_formula"] !=  df["href"]].shape

(60, 24)

Llegados a este punto nos vemos obligados a analizarlo como un proyecto independiente para saber los enlaces que nos puedan aportar informacion  [aqui](analisi_href.ipynb). Al final de este analisis se consigue un listado de url operativa.

In [68]:
# cargamos el archivo href
df_href = pd.read_csv("../data/output/href_limpio.csv",encoding = "ISO-8859-1")
df_href.drop(columns="href_response",inplace=True)
df_href

href
0     http://sharkattackfile.net/spreadsheets/pdf_di...
1     http://sharkattackfile.net/spreadsheets/pdf_di...
2     http://sharkattackfile.net/spreadsheets/pdf_di...
3     http://sharkattackfile.net/spreadsheets/pdf_di...
4     http://sharkattackfile.net/spreadsheets/pdf_di...
...                                                 ...
6102  http://sharkattackfile.net/spreadsheets/pdf_di...
6103  http://sharkattackfile.net/spreadsheets/pdf_di...
6104  http://sharkattackfile.net/spreadsheets/pdf_di...
6105  http://sharkattackfile.net/spreadsheets/pdf_di...
6106  http://sharkattackfile.net/spreadsheets/pdf_di...

[6107 rows x 1 columns]

In [69]:
# dropeamos la columna href_formula y href de df para hacer un merge con df_href
df.drop(columns=["href_formula","href"],inplace=True)
df.head()

case_number         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal_(y/n)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            time          species          investigator_or_source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf case_number.1 case_number.2  original_order  \
0      2018.06.25-Wolfe.pdf    2018.06.25    2018.06.25          6303.0   
1    2018.06.18-McNeely.pdf    2018.06.18    2018.06.18          6302.0   
2     2018.06.09-Denges.pdf    2018.06.09    2018.06.09          6301.0   
3  2018.06.08-Arrawarra.pdf    2018.06.08    2018.06.08          6300.0   
4      2018.06.04-Ramos.pdf    2018.06.04    2018.06.04          6299.0   

  unnamed:_22 unnamed:_23  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  
3         NaN         NaN  
4         NaN         NaN

In [71]:
# hacemos un merge con df_href 
df = df.merge(df_href, left_index=True, right_index=True)
df.head()

case_number         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal_(y/n)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            time          species          investigator_or_source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf case_number.1 case_number.2  original_order  \
0      2018.06.25-Wolfe.pdf    2018.06.25    2018.06.25          6303.0   
1    2018.06.18-McNeely.pdf    2018.06.18    2018.06.18          6302.0   
2     2018.06.09-Denges.pdf    2018.06.09    2018.06.09          6301.0   
3  2018.06.08-Arrawarra.pdf    2018.06.08    2018.06.08          6300.0   
4      2018.06.04-Ramos.pdf    2018.06.04    2018.06.04          6299.0   

  unnamed:_22 unnamed:_23                                               href  
0         NaN         NaN  http://sharkattackfile.net/spreadsheets/pdf_di...  
1         NaN         NaN  http://sharkattackfile.net/spreadsheets/pdf_di...  
2         NaN         NaN  http://sharkattackfile.net/spreadsheets/pdf_di...  
3         NaN         NaN  http://sharkattackfile.net/spreadsheets/pdf_di...  
4         NaN         NaN  http://sharkattackfile.net/spreadsheets/pdf_di...

#### COLUMNAS

In [72]:
# miramos los valores nulos ordenados de mayor a menor
df.isnull().sum().sort_values(ascending=False)

unnamed:_22               6106
unnamed:_23               6105
time                      3172
species                   2693
age                       2653
sex                        546
fatal_(y/n)                528
activity                   500
location                   484
area                       397
name                       198
country                     38
injury                      20
investigator_or_source      15
type                         4
year                         2
case_number                  1
case_number.1                0
case_number.2                0
original_order               0
pdf                          0
date                         0
href                         0
dtype: int64

In [73]:
# miramos los valores nulos en porcentaje valores ordenados de mayor a menor
df.isnull().sum().sort_values(ascending=False)/df.shape[0]*100


unnamed:_22               99.983625
unnamed:_23               99.967251
time                      51.940396
species                   44.096938
age                       43.441952
sex                        8.940560
fatal_(y/n)                8.645816
activity                   8.187326
location                   7.925332
area                       6.500737
name                       3.242181
country                    0.622237
injury                     0.327493
investigator_or_source     0.245620
type                       0.065499
year                       0.032749
case_number                0.016375
case_number.1              0.000000
case_number.2              0.000000
original_order             0.000000
pdf                        0.000000
date                       0.000000
href                       0.000000
dtype: float64

> Resmunen: analisis de nulos 

In [74]:
# generamos un dataframe con una columna con los valores nulos en valores absolutos y otra columna con porcentaje
df_nulls = pd.DataFrame(df.isnull().sum().sort_values(ascending=False), columns=["valores_nulos"])
df_nulls["porcentaje"] = df.isnull().sum().sort_values(ascending=False)/df.shape[0]*100
df_nulls

valores_nulos  porcentaje
unnamed:_22                      6106   99.983625
unnamed:_23                      6105   99.967251
time                             3172   51.940396
species                          2693   44.096938
age                              2653   43.441952
sex                               546    8.940560
fatal_(y/n)                       528    8.645816
activity                          500    8.187326
location                          484    7.925332
area                              397    6.500737
name                              198    3.242181
country                            38    0.622237
injury                             20    0.327493
investigator_or_source             15    0.245620
type                                4    0.065499
year                                2    0.032749
case_number                         1    0.016375
case_number.1                       0    0.000000
case_number.2                       0    0.000000
original_order                      0    0.000000
pdf                                 0    0.000000
date                                0    0.000000
href                                0    0.000000

In [75]:
#generamos una funcion a partir de la ultima celda 
def porcentaje_nulos(df):
    """funcion para generar un dataframe con los 
        valores nulos en valores absolutos y otra 
        columna con porcentaje.

    Args:
        df (PandasDataframe): dataframe a analizar.

    Returns:
        PandasDataframe: dataframe con los valores nulos en  valores 
                         absolutos y otra columna con porcentaje.
    """
    df_nulls = pd.DataFrame(df.isnull().sum().sort_values(ascending=False), columns=["valores_nulos"])
    df_nulls["porcentaje"] = df.isnull().sum().sort_values(ascending=False)/df.shape[0]*100
    return df_nulls

In [76]:
porcentaje_nulos(df)

valores_nulos  porcentaje
unnamed:_22                      6106   99.983625
unnamed:_23                      6105   99.967251
time                             3172   51.940396
species                          2693   44.096938
age                              2653   43.441952
sex                               546    8.940560
fatal_(y/n)                       528    8.645816
activity                          500    8.187326
location                          484    7.925332
area                              397    6.500737
name                              198    3.242181
country                            38    0.622237
injury                             20    0.327493
investigator_or_source             15    0.245620
type                                4    0.065499
year                                2    0.032749
case_number                         1    0.016375
case_number.1                       0    0.000000
case_number.2                       0    0.000000
original_order                      0    0.000000
pdf                                 0    0.000000
date                                0    0.000000
href                                0    0.000000

### 3.3.2.Tratamiento de nulos 

En este ejercicio trataremos las columnas en funcion de la proporcion de nulos y los insights del apartado de Filas

---
Vemos como tenemos muchos nulos en `unnamed:_22` y `unnamed:_23` procedemos a eliminar esas columnas. 

In [77]:
# eliminamos las columnas unmnamed:22 y unnamed:23 que no aportan informacion
df.drop(["unnamed:_22","unnamed:_23"], axis=1, inplace=True)

In [78]:
# confirmamos que se han eliminado
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'case_number.1',
       'case_number.2', 'original_order', 'href'],
      dtype='object')

---
Tratamiento de la columnas de `time`

In [79]:
df.head(3)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid     USA      Hawaii   

                         location  activity             name sex age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe   F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely    F  11   
2                    Habush, Oahu   Surfing      John Denges   M  48   

                                              injury fatal_(y/n)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   

            time      species          investigator_or_source  \
0          18h00  White shark                R. Collier, GSAF   
1  14h00  -15h00          NaN  K.McMurray, TrackingSharks.com   
2          07h45          NaN  K.McMurray, TrackingSharks.com   

                      pdf case_number.1 case_number.2  original_order  \
0    2018.06.25-Wolfe.pdf    2018.06.25    2018.06.25          6303.0   
1  2018.06.18-McNeely.pdf    2018.06.18    2018.06.18          6302.0   
2   2018.06.09-Denges.pdf    2018.06.09    2018.06.09          6301.0   

                                                href  
0  http://sharkattackfile.net/spreadsheets/pdf_di...  
1  http://sharkattackfile.net/spreadsheets/pdf_di...  
2  http://sharkattackfile.net/spreadsheets/pdf_di...

In [80]:
df['time'].value_counts(dropna=False)

time
NaN            3172
Afternoon       183
11h00           128
Morning         120
12h00           109
               ... 
14h37             1
10h07             1
13h53             1
13h23             1
19h00-20h00       1
Name: count, Length: 367, dtype: int64

In [81]:
# miramos los unique de la columna time 
df["time"].unique()

array(['18h00', '14h00  -15h00', '07h45', nan, 'Late afternoon', '17h00',
       '14h00', 'Morning', '15h00', '08h15', '11h00', '10h30', '10h40',
       '16h50', '07h00', '09h30', 'Afternoon', '21h50', '09h40', '08h00',
       '17h35', '15h30', '07h30', '19h00, Dusk', 'Night', '16h00',
       '15h01', '12h00', '13h45', '23h30', '09h00', '14h30', '18h30',
       '12h30', '16h30', '18h45', '06h00', '10h00', '10h44', '13h19',
       'Midday', '13h30', '10h45', '11h20', '11h45', '19h30', '08h30',
       '15h45', 'Shortly before 12h00', '17h34', '17h10', '11h15',
       '08h50', '17h45', '13h00', '10h20', '13h20', '02h00', '09h50',
       '11h30', '17h30', '9h00', '10h43', 'After noon', '15h15', '15h40',
       '19h05', '1300', '14h30 / 15h30', '22h00', '16h20', '14h34',
       '15h25', '14h55', '17h46', 'Morning ', '15h49', '19h00',
       'Midnight', '09h30 / 10h00', '10h15', '18h15', '04h00', '14h50',
       '13h50', '19h20', '10h25', '10h45-11h15', '16h45', '15h52',
       '06h15', '14h

In [82]:
# alamacenamos los valores que contengan un numero en lista_num y el resto en data_qa
lista_num = []
data_qa = []
for x in df["time"]:
    if re.search("\d",str(x)):
        lista_num.append(x)
    else:
        data_qa.append(x)


In [83]:
# miramos los valores que contienen un numero
print(lista_num)

['18h00', '14h00  -15h00', '07h45', '17h00', '14h00', '15h00', '08h15', '11h00', '10h30', '10h40', '16h50', '07h00', '09h30', '21h50', '09h40', '08h00', '09h30', '17h00', '18h00', '15h00', '09h30', '17h35', '15h30', '07h30', '19h00, Dusk', '16h00', '18h00', '08h00', '15h30', '15h01', '18h00', '12h00', '13h45', '16h00', '23h30', '09h00', '15h00', '17h00', '08h00', '09h00', '14h30', '18h30', '12h30', '18h00', '15h30', '14h00', '16h30', '18h45', '15h00', '06h00', '10h00', '10h44', '10h30', '13h19', '10h00', '15h00', '13h30', '23h30', '10h45', '18h30', '15h00', '17h00', '12h30', '11h20', '07h45', '18h30', '11h45', '11h00', '15h30', '15h00', '19h30', '10h00', '12h00', '08h30', '15h45', '10h00', 'Shortly before 12h00', '17h00', '14h30', '17h34', '17h10', '11h00', '11h15', '14h00', '16h00', '08h50', '16h00', '09h00', '13h30', '13h45', '16h00', '17h45', '17h00', '13h00', '15h00', '10h00', '18h30', '14h30', '09h30', '10h20', '13h30', '13h20', '14h00', '12h00', '13h30', '07h30', '12h30', '18h00'

In [84]:
# miramos los valores unicos de la lista_num
print(set(lista_num))

{'10h40', '08h10', '12h10', '12h35', '09h57', '10h44', '11h06', '15h50', '14h16', '17h00 Sunset', '10h43', '17h58', '12h02', '17h00', '18h30', '13h51', '18h30?', '10h10', '2 hours after Opperman', '18h25', '19h00, Dusk', '12h40', '07h45', '18h30 (Sunset)', '18h10', '11h57', '22h30', '07h15', '13h53', '08h50', '05h45', '06h00 -- 07h00', '14h37', '12h45 / 13h45', '01h00', '08h40', '15h45', '18h50', '08h00', 'Before 10h30', '14h10', '12h39', 'Just after 12h00', '18h40', '14h15', '16h30 or 18h00', '17h35', '16h20', '11h15', '12h05', '11h55', '02h45', '10h30', '07h56', 'Between 11h00 & 12h00', '15h35', 'Possibly same incident as 2000.08.21', '18h05', '10h45-11h15', 'Ship aban-doned at 03h10', '03h30', '23h30', '19h45', '06h30', '15h20', '07h00 - 08h00', '11h25', '10h35', '10h16', '12h25', '1600', '20h00', '11h51', '11h48', '12h55', '16h14', '17h30', '09h15', '>14h30', '13h19', '10h00', '11h45', '01h30', '15h05', '17h00-18h00', '09h30', '13h37', '08h15', '17h11', '14h40', '22h00', '10h27', '

In [85]:
x = '17h11'
pattern="^\d{2}|[A-Z][a-z]*" # este patron de regex las horas. d{2} dos digitos, | significa o, [A-Z] una letra mayuscula, [a-z]* cero o mas letras minusculas
res = re.findall(pattern, x)
res

['17']

In [86]:
"".join(res)

'17'

In [87]:
# miramos los valores que no contienen un numero
print(data_qa)

[nan, nan, nan, 'Late afternoon', nan, 'Morning', nan, nan, nan, 'Afternoon', nan, nan, nan, nan, nan, nan, 'Night', nan, nan, nan, nan, 'Late afternoon', nan, nan, nan, nan, nan, nan, 'Late afternoon', nan, nan, 'Afternoon', 'Afternoon', nan, 'Morning', 'Morning', nan, 'Afternoon', 'Night', nan, nan, 'Morning', nan, nan, 'Midday', nan, nan, 'Afternoon', nan, nan, nan, 'Afternoon', nan, nan, nan, nan, 'Morning', nan, nan, nan, nan, nan, nan, 'Morning', nan, 'Morning', nan, nan, 'Afternoon', nan, nan, 'Afternoon', nan, nan, 'Morning', 'Morning', 'Afternoon', nan, nan, nan, 'Morning', nan, 'Morning', nan, 'Afternoon', 'Afternoon', 'Afternoon', nan, 'After noon', nan, nan, nan, 'Late afternoon', 'Late afternoon', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 'After noon', nan, nan, nan, nan, nan, nan, nan, nan, 'Morning', nan, 'Afternoon', 'Morning ', 'Afternoon', 'Morning', 'Morning', nan, nan, nan, nan, nan, nan, 'Morning', 'Afternoon', 'Afternoon', 'Midnight', nan, nan, 'Late after

> Resumen: funcion de tiempo

In [88]:
def daytime (x, pattern):
    """ funcion que devuelve el periodo del dia en el que se produjo el ataque

    Args:
        x (str): string que contiene la fecha del ataque 
        pattern (_type_): patron de busqueda

    Returns:
        np.nan: devuelve un nan en el caso de que no se encuentre el patron
        str : devuelve el estado de la fecha del ataque (Morning, Afternoon, Evening, Night)
    """
    import re
    res = re.findall(pattern, x)
    res="".join(res)
    if res in ["06","07","08","09","10","11", "Morning"]:
        return "Morning"
    elif res in ["12","13","14","15","16","17", "Afternoon"]:
        return "Afternoon"
    elif res in ["18","19","20","21","22","23", "Evening"]:
        return "Evening"
    elif res in ["24","01","02","03","04","05", "Night"]:
        return "Night"
    else:
        return x

In [89]:
pattern="^\d{2}|[A-Z][a-z]*"

df["time"]=df["time"].astype(str)
df['time'] = df['time'].apply(lambda x: daytime(x, pattern))

- Comprobacion 

In [90]:
df['time'].value_counts(dropna=False).head()

time
nan          3172
Afternoon    1495
Morning       888
Evening       260
Night         100
Name: count, dtype: int64

---
Tratamiento de la columnas de `species`

In [91]:
df["species"].value_counts(dropna=False).head(15)

species
NaN                                                   2693
White shark                                            161
Shark involvement prior to death was not confirmed     105
Invalid                                                100
Shark involvement not confirmed                         88
Tiger shark                                             70
Shark involvement prior to death unconfirmed            64
Bull shark                                              48
4' shark                                                40
6' shark                                                39
Questionable incident                                   35
1.8 m [6'] shark                                        33
Questionable                                            32
1.5 m [5'] shark                                        32
3' shark                                                26
Name: count, dtype: int64

Esta información me parece interesante, sin embargo, los datos están muy mal registrados. Hay muchas frases largas que significan lo mismo.
Regex:
- Tiburones definidos por nombres (mayúscula + tiburón):` ([A-Z][a-z]*\sshark) `
- Tiburones definidos por su longitud (en yardas):` \d+']\sshark|\d+'\sshark `
- No confirmado: (no confirmado)|(Inválido)|(Cuestionable)|(sin confirmar)


In [92]:
df["species"]=df["species"].astype(str)
df["species"]

0           White shark
1                   nan
2                   nan
3             2 m shark
4       Tiger shark, 3m
             ...       
6102                nan
6103                nan
6104                nan
6105                nan
6106                nan
Name: species, Length: 6107, dtype: object

In [93]:
list(df["species"])

['White shark',
 'nan',
 'nan',
 '2 m shark',
 'Tiger shark, 3m',
 'nan',
 'Tiger shark',
 "Lemon shark, 3'",
 "Bull shark, 6'",
 'nan',
 'Grey reef shark',
 'nan',
 'nan',
 'Invalid incident',
 'Tawny nurse shark, 2m',
 'nan',
 'Shark involvement not confirmed',
 'nan',
 'Tiger shark',
 'nan',
 'Questionable',
 '3 m shark',
 'White shark, 3.5 m',
 'Tiger shark',
 'White shark, 2.5 m',
 "6' shark",
 'Juvenile bull shark',
 'nan',
 'nan',
 'nan',
 'Tiger shark',
 'Shark involvement not confirmed',
 '2 m shark',
 'Bull shark',
 'White shark',
 "Tiger shark, 12'",
 'Wobbegong shark',
 '3.5 m shark',
 '1.8 m shark',
 'Blacktip shark',
 'Juvenile white shark,  2.7 to 3.2 m',
 'nan',
 'Bull shark, 2 m',
 'Possibly a wobbegong',
 'Injury believed caused by an eel, not a shark',
 'Galapagos shark?',
 '2m shark',
 'Bull shark',
 'Bull shark, 3 m ',
 'Grey reef shark. 2 m',
 'White shark, 3.5 m',
 'small shark',
 'Wobbegong shark?',
 'Tiger shark',
 'nan',
 'Juvenile nurse shark',
 "Nurse shark.

In [94]:
#[A-Z][a-z]*\sshark|\d+'] indica que queremos buscar una palabra que empiece por mayuscula y acabe por shark o un numero seguido de ' y shark, 
#\sshark| indica que queremos buscar una palabra que empiece por un numero seguido de ' y shark , 
#d+'\sshark indica que queremos buscar un numero seguido de ' y shark
pattern= "[A-Z][a-z]*\sshark|\d+']\sshark|\d+'\sshark" 
x = "Tiger shark, 3m"

res = re.findall(pattern, x)
res

['Tiger shark']

In [95]:
"".join(res)

'Tiger shark'

In [ ]:
pattern= "[A-Z][a-z]*\sshark|\d+']\sshark|\d+'\sshark" 
x = "Mako shark, 1.5 m [5'] "

res = re.findall(pattern, x)
res=res.replace("]", "")
res

In [97]:
def regeshark (x, pattern):
    """ funcion que devuelve el tipo de tiburon que ha atacado usando regex y el patron de busqueda

    Args:
        x (str): string que contiene el tipo de tiburon
        pattern (str): patron de busqueda

    Returns:
        np.nan: en caso en el que no se encuentre el patron devuelve un nan
        str: devuelve el tipo de tiburon
    """
    res = re.findall(pattern, x)
    res="".join(res)
    res=res.replace("]", "")
    if "shark" in res:
        return res
    else:
        return np.nan

In [98]:
pattern= "[A-Z][a-z]*\sshark|\d+']\sshark|\d+'\sshark"
df['species_sorted'] = df['species'].apply(lambda x: regeshark(x, pattern))

In [99]:
df['species_sorted'].value_counts(dropna=False).head(20)

species_sorted
NaN                  4214
White shark           430
Tiger shark           233
5' shark              131
Bull shark            125
6' shark              101
4' shark               97
8' shark               51
Nurse shark            47
Wobbegong shark        46
Mako shark             44
3' shark               43
Raggedtooth shark      43
10' shark              41
12' shark              39
7' shark               36
Blacktip shark         33
Lemon shark            32
Blue shark             29
Zambesi shark          29
Name: count, dtype: int64

Parece que ya tenemos una columna de especies limpia

---
Tratamiento de la columnas de `age`

En la columna `age` aparecen ciertos valores como strings por lo que necesitamos transformarlos en integer. 

Para los valores nulos podremos estimarlos calculando la media y la std para posteriormente inferir de manera aleatoria valores sin peder informacion. 

In [100]:
df['age'].value_counts(dropna=False).head()

age
NaN    2653
17      154
18      150
20      140
19      140
Name: count, dtype: int64

In [101]:
df['age'].unique()

array(['57', '11', '48', nan, '18', '52', '15', '12', '32', '10', '21',
       '34', '30', '60', '33', '29', '54', '41', '37', '56', '19', '25',
       '69', '38', '55', '35', '46', '45', '14', '40s', '28', '20', '24',
       '26', '49', '22', '7', '31', '17', '40', '13', '42', '3', '8',
       '50', '16', '82', '73', '20s', '68', '51', '39', '58', 'Teen',
       '47', '61', '65', '36', '66', '43', '60s', '9', '72', '59', '6',
       '27', '64', '23', '71', '44', '62', '63', '70', '18 months', '53',
       '30s', '50s', 'teen', '77', '74', '28 & 26', '5', '86', '18 or 20',
       '12 or 13', '46 & 34', '28, 23 & 30', 'Teens', '36 & 26',
       '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75',
       '33 or 37', 'mid-30s', '23 & 20', ' 30', '7      &    31', ' 28',
       '20?', "60's", '32 & 30', '16 to 18', '87', '67', 'Elderly',
       'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26', '20 ', '>50',
       '18 to 22', 'adult', '9 & 12', '? & 19', '9 months', '25 to 35',
  

In [108]:
# fucnion que crea lista de numeros en un string
def age(x):
    """ funcion que crea una lista de numeros en un string

    Args:
        x (str): string a analizar

    Returns:
        list: lista de numeros en un string
    """
    if type(x) != str:
        return np.nan
    else:
        return re.findall("\d+", x)


In [109]:
df['age']=df['age'].astype(str)
df['age2']=df['age'].apply(lambda x: age(x))

In [110]:
df.head(3)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid     USA      Hawaii   

                         location  activity             name sex age  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe   F  57   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely    F  11   
2                    Habush, Oahu   Surfing      John Denges   M  48   

                                              injury fatal_(y/n)       time  \
0  No injury to occupant, outrigger canoe and pad...           N    Evening   
1                         Minor injury to left thigh           N  Afternoon   
2       Injury to left lower leg from surfboard skeg           N    Morning   

       species          investigator_or_source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          nan  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   
2          nan  K.McMurray, TrackingSharks.com   2018.06.09-Denges.pdf   

  case_number.1 case_number.2  original_order  \
0    2018.06.25    2018.06.25          6303.0   
1    2018.06.18    2018.06.18          6302.0   
2    2018.06.09    2018.06.09          6301.0   

                                                href species_sorted  age2  
0  http://sharkattackfile.net/spreadsheets/pdf_di...    White shark  [57]  
1  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN  [11]  
2  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN  [48]

Se observa que tenemos varios años, en esos casos significa que ese registro tiene varias victimas. Una posible solucion seria la de aplicar un `explode` para esos registros.

In [122]:



# generamos una columna para aplicar un explode y mantenemos la columna age2 para comprobar 
df['age3'] = df['age2']
# explode de una columna age3
df  = df.explode("age3")
df

case_number                  date    year          type    country  \
0       2018.06.25           25-Jun-2018  2018.0       Boating        USA   
1       2018.06.18           18-Jun-2018  2018.0    Unprovoked        USA   
2       2018.06.09           09-Jun-2018  2018.0       Invalid        USA   
3       2018.06.08           08-Jun-2018  2018.0    Unprovoked  AUSTRALIA   
4       2018.06.04           04-Jun-2018  2018.0      Provoked     MEXICO   
...            ...                   ...     ...           ...        ...   
6102    1826.11.00          Ca. Nov-1826  1826.0    Unprovoked      GHANA   
6103    1826.08.28           28-Aug-1826  1826.0  Sea Disaster       CUBA   
6104  1826.08.15.R  Reported 15-Aug-1826  1826.0    Unprovoked    ENGLAND   
6105  1826.05.20.R  Reported 20-May-1826  1826.0    Unprovoked    JAMAICA   
6106  1825.00.00.b                 1820s  1825.0    Unprovoked  AUSTRALIA   

                 area                        location  \
0          California     Oceanside, San Diego County   
1             Georgia  St. Simon Island, Glynn County   
2              Hawaii                    Habush, Oahu   
3     New South Wales              Arrawarra Headland   
4              Colima                        La Ticla   
...               ...                             ...   
6102       Cape Coast                             NaN   
6103              NaN                             NaN   
6104       Cumberland                      Whitehaven   
6105              NaN                             NaN   
6106         Tasmania   Sweet Water Point, Pitt Water   

                              activity                      name  sex  age  \
0                             Paddling               Julie Wolfe    F   57   
1                             Standing           Adyson McNeely     F   11   
2                              Surfing               John Denges    M   48   
3                              Surfing                      male    M  nan   
4                          Free diving            Gustavo Ramos     M  nan   
...                                ...                       ...  ...  ...   
6102                           Bathing  a seaman from HM Redwing    M  nan   
6103  HBM Magpie foundered in a squall   Lieutenant Edward Smith    M  nan   
6104                           Bathing                     child  NaN  nan   
6105                  Jumped overboard             Thomas Loring    M  nan   
6106                          Swimming    "Amphibious Jack" male    M  nan   

                                                 injury fatal_(y/n)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6102                                              FATAL           Y   
6103                                             FATAL            Y   
6104                                              FATAL         NaN   
6105                                              FATAL           Y   
6106                                              FATAL           Y   

           time          species  \
0       Evening      White shark   
1     Afternoon              nan   
2       Morning              nan   
3           nan        2 m shark   
4           nan  Tiger shark, 3m   
...         ...              ...   
6102        nan              nan   
6103        nan              nan   
6104        nan              nan   
6105    Evening              nan   
6106        nan              nan   

                                 investigator_or_source  \
0                                      R. Collier, GSAF   
1                        K.McMurray, TrackingSharks.com 

In [123]:
# ahora podemos rellenar los valores nulos de la columna age3 haciendo inferencia mediante la normal tras calcular la media y std de la columna age3

# calculamos la media y std de la columna age3
media = df["age3"].astype(float).mean()
std = df["age3"].astype(float).std()


In [125]:
# importamos las librerias para generar una normal 
import scipy.stats as stats


In [126]:

# generamos un objeto para crear una distribucion normal con la media y std de la columna age3
dist = stats.norm(media, std)


In [130]:
# rellenamos los valores nulos de la columna age3 con valores aleatorios generados por la distribucion normal
df["age3"] = df["age3"].apply(lambda x: dist.rvs() if pd.isnull(x) else x)

df

case_number                  date    year          type    country  \
0       2018.06.25           25-Jun-2018  2018.0       Boating        USA   
1       2018.06.18           18-Jun-2018  2018.0    Unprovoked        USA   
2       2018.06.09           09-Jun-2018  2018.0       Invalid        USA   
3       2018.06.08           08-Jun-2018  2018.0    Unprovoked  AUSTRALIA   
4       2018.06.04           04-Jun-2018  2018.0      Provoked     MEXICO   
...            ...                   ...     ...           ...        ...   
6102    1826.11.00          Ca. Nov-1826  1826.0    Unprovoked      GHANA   
6103    1826.08.28           28-Aug-1826  1826.0  Sea Disaster       CUBA   
6104  1826.08.15.R  Reported 15-Aug-1826  1826.0    Unprovoked    ENGLAND   
6105  1826.05.20.R  Reported 20-May-1826  1826.0    Unprovoked    JAMAICA   
6106  1825.00.00.b                 1820s  1825.0    Unprovoked  AUSTRALIA   

                 area                        location  \
0          California     Oceanside, San Diego County   
1             Georgia  St. Simon Island, Glynn County   
2              Hawaii                    Habush, Oahu   
3     New South Wales              Arrawarra Headland   
4              Colima                        La Ticla   
...               ...                             ...   
6102       Cape Coast                             NaN   
6103              NaN                             NaN   
6104       Cumberland                      Whitehaven   
6105              NaN                             NaN   
6106         Tasmania   Sweet Water Point, Pitt Water   

                              activity                      name  sex  age  \
0                             Paddling               Julie Wolfe    F   57   
1                             Standing           Adyson McNeely     F   11   
2                              Surfing               John Denges    M   48   
3                              Surfing                      male    M  nan   
4                          Free diving            Gustavo Ramos     M  nan   
...                                ...                       ...  ...  ...   
6102                           Bathing  a seaman from HM Redwing    M  nan   
6103  HBM Magpie foundered in a squall   Lieutenant Edward Smith    M  nan   
6104                           Bathing                     child  NaN  nan   
6105                  Jumped overboard             Thomas Loring    M  nan   
6106                          Swimming    "Amphibious Jack" male    M  nan   

                                                 injury fatal_(y/n)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6102                                              FATAL           Y   
6103                                             FATAL            Y   
6104                                              FATAL         NaN   
6105                                              FATAL           Y   
6106                                              FATAL           Y   

           time          species  \
0       Evening      White shark   
1     Afternoon              nan   
2       Morning              nan   
3           nan        2 m shark   
4           nan  Tiger shark, 3m   
...         ...              ...   
6102        nan              nan   
6103        nan              nan   
6104        nan              nan   
6105    Evening              nan   
6106        nan              nan   

                                 investigator_or_source  \
0                                      R. Collier, GSAF   
1                        K.McMurray, TrackingSharks.com 

In [116]:
df.drop(columns=["age3"], inplace=True)

Voy a hacer lo mismo con la columna Fecha, hay 5 columnas con información sobre la fecha, es decir,  "Case Number",	"date", "Year", "Case Number.1" and "Case Number.2".  El año puede dar alguna información, así que me quedo con esta columna, de las otras 4 columnas sólo me interesa coger el mes de la columna "date", quiero ver si hay una correlación entre la época del año y los ataques de tiburón.

In [131]:
pattern="[A-Z][a-z]+"


df["date"]=df["date"].astype(str)

df['month_attack'] = df['date'].apply(lambda x: tn.monthattack(x, pattern))

In [132]:
df['month_attack'].sample(10)

1589    May
4990    May
441     Mar
1103    Jul
1125    May
1778    Apr
4031    NaN
3020    Jul
5044    May
3960    Oct
Name: month_attack, dtype: object

Ahora tengo una columna con los meses en los que se produjeron los atentados.

Las columnas "Activity" y "Type" también parecen interesantes, pero ya me parecen bastante limpias, así que las mantengo como recibidas.

In [133]:
df['activity'].value_counts(dropna=False).head(10)

activity
Surfing         971
Swimming        842
NaN             502
Fishing         423
Spearfishing    333
Bathing         155
Wading          146
Diving          115
Standing         99
Snorkeling       88
Name: count, dtype: int64

In [134]:
df['type'].value_counts(dropna=False).head(10)

type
Unprovoked      4454
Provoked         568
Invalid          549
Sea Disaster     230
Boating          208
Boat             134
NaN                4
Questionable       2
Boatomg            1
Name: count, dtype: int64

La columna "Time" mezcla horas y datos diurnos, creo que aportará más información si la ordeno por mañana, tarde, noche y noche.

In [135]:
df['time'].value_counts(dropna=False).head(20)

time
nan                3196
Afternoon          1505
Morning             895
Evening             260
Night               101
Late afternoon       34
Dusk                 15
A.M.                 12
P.M.                 12
Early morning        11
Midday               10
Early afternoon       8
--                    5
Midnight              4
Sunset                4
Late night            2
Dark                  2
Late afternon         2
                      2
AM                    2
Name: count, dtype: int64

In [136]:
pattern="^\d{2}|[A-Z][a-z]*"

df["time"]=df["time"].astype(str)
df['time'] = df['time'].apply(lambda x: sp.daytime(x, pattern))

In [137]:
df['time'].value_counts(dropna=False).head()

time
NaN          3389
Afternoon    1505
Morning       895
Evening       260
Night         101
Name: count, dtype: int64

La columna de lesiones parece limpia, no hay mucha información que obtener de ella, pero la mantendré.

In [139]:
df['fatal_(y/n)'].value_counts(dropna=False).head(50)

fatal_(y/n)
N          4237
Y          1300
NaN         540
UNKNOWN      63
 N            7
M             1
2017          1
N             1
Name: count, dtype: int64

In [140]:
pattern='^N|^Y'

df['fatal_(y/n)']=df['fatal_(y/n)'].astype(str)
df['fatal_(y/n)']=df['fatal_(y/n)'].apply(lambda x: sp.fatal(x, pattern))

In [141]:
df['fatal_(y/n)'].value_counts(dropna=False).head()

fatal_(y/n)
N      4238
Y      1300
NaN     612
Name: count, dtype: int64

En la columna Sexo hay algunos valores erróneos, así que vamos a limpiarla también

In [142]:
df['sex'].value_counts(dropna=False).head()

sex
M      4966
F       628
NaN     551
M         2
lli       1
Name: count, dtype: int64

In [144]:
pattern='[M,F]'
    
df['sex']=df['sex'].astype(str)
df['sex']=df['sex'].apply(lambda x: sp.sex(x, pattern))

In [145]:
df['sex'].value_counts(dropna=False).head()

sex
M      4968
F       628
NaN     554
Name: count, dtype: int64

In [146]:
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'case_number.1',
       'case_number.2', 'original_order', 'href', 'species_sorted', 'age2',
       'age3', 'month_attack'],
      dtype='object')

In [147]:
df.head()

case_number         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  nan   
4                        La Ticla  Free diving   Gustavo Ramos    M  nan   

                                              injury fatal_(y/n)       time  \
0  No injury to occupant, outrigger canoe and pad...           N    Evening   
1                         Minor injury to left thigh           N  Afternoon   
2       Injury to left lower leg from surfboard skeg           N    Morning   
3                          Minor injury to lower leg           N        NaN   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N        NaN   

           species          investigator_or_source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              nan  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              nan  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

  case_number.1 case_number.2  original_order  \
0    2018.06.25    2018.06.25          6303.0   
1    2018.06.18    2018.06.18          6302.0   
2    2018.06.09    2018.06.09          6301.0   
3    2018.06.08    2018.06.08          6300.0   
4    2018.06.04    2018.06.04          6299.0   

                                                href species_sorted  age2  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    White shark  [57]   
1  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN  [11]   
2  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN  [48]   
3  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN    []   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    Tiger shark    []   

        age3 month_attack  
0         57          Jun  
1         11          Jun  
2         48          Jun  
3  38.590376          Jun  
4  36.174184          Jun

Ahora, todas las columnas que necesito están limpias y ordenadas, sólo tengo que crear un nuevo conjunto de datos con las columnas que quiero e importarlo.

In [148]:
df.columns


Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'case_number.1',
       'case_number.2', 'original_order', 'href', 'species_sorted', 'age2',
       'age3', 'month_attack'],
      dtype='object')

In [149]:
# dropeamos columnas que no vamos a usar
df.drop(columns=['age2','case_number.1',
                'case_number.2', 'original_order',
                'investigator_or_source', 'pdf', 
                'age',"name", "case_number"], inplace=True)

In [150]:
df

date    year          type    country             area  \
0              25-Jun-2018  2018.0       Boating        USA       California   
1              18-Jun-2018  2018.0    Unprovoked        USA          Georgia   
2              09-Jun-2018  2018.0       Invalid        USA           Hawaii   
3              08-Jun-2018  2018.0    Unprovoked  AUSTRALIA  New South Wales   
4              04-Jun-2018  2018.0      Provoked     MEXICO           Colima   
...                    ...     ...           ...        ...              ...   
6102          Ca. Nov-1826  1826.0    Unprovoked      GHANA       Cape Coast   
6103           28-Aug-1826  1826.0  Sea Disaster       CUBA              NaN   
6104  Reported 15-Aug-1826  1826.0    Unprovoked    ENGLAND       Cumberland   
6105  Reported 20-May-1826  1826.0    Unprovoked    JAMAICA              NaN   
6106                 1820s  1825.0    Unprovoked  AUSTRALIA         Tasmania   

                            location                          activity  sex  \
0        Oceanside, San Diego County                          Paddling    F   
1     St. Simon Island, Glynn County                          Standing    F   
2                       Habush, Oahu                           Surfing    M   
3                 Arrawarra Headland                           Surfing    M   
4                           La Ticla                       Free diving    M   
...                              ...                               ...  ...   
6102                             NaN                           Bathing    M   
6103                             NaN  HBM Magpie foundered in a squall    M   
6104                      Whitehaven                           Bathing  NaN   
6105                             NaN                  Jumped overboard    M   
6106   Sweet Water Point, Pitt Water                          Swimming    M   

                                                 injury fatal_(y/n)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6102                                              FATAL           Y   
6103                                             FATAL            Y   
6104                                              FATAL         NaN   
6105                                              FATAL           Y   
6106                                              FATAL           Y   

           time          species  \
0       Evening      White shark   
1     Afternoon              nan   
2       Morning              nan   
3           NaN        2 m shark   
4           NaN  Tiger shark, 3m   
...         ...              ...   
6102        NaN              nan   
6103        NaN              nan   
6104        NaN              nan   
6105    Evening              nan   
6106        NaN              nan   

                                                   href species_sorted  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...    White shark   
1     http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
2     http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
3     http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
4     http://sharkattackfile.net/spreadsheets/pdf_di...    Tiger shark   
...                                                 ...            ...   
6102  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
6103  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
6104  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
6105  http://sharkattackfile.net/spreadsheets/pdf_di...            NaN   
6106  http:/

In [151]:
# guardamos el dataframe en un csv
df.to_csv("../data/output/attacks_clean.csv", index=False)
